contracts/test/PDPService.t.sol

```diff
@@ -313,6 +316,133 @@ contract PDPServiceProofSetMutateTest is Test {
+        for (uint i = 0; i < rootCount; i++) {
+            // Generate different size trees up to max leaves.
+            leafCounts[i] = uint256(sha256(abi.encode(i))) % 64;
+            trees[i] = makeTree(leafCounts[i]);
+            roots[i] = makeRoot(trees[i], leafCounts[i]);
+        }
+
+        // Create new proof set and add roots.
+        uint256 setId = pdpService.createProofSet();
+        pdpService.addRoots(setId, roots);
+
+        // Advance chain until challenge epoch.
+        uint256 challengeEpoch = pdpService.getNextChallengeEpoch(setId);
+        vm.roll(challengeEpoch);
+
+        // Build a proof with multiple challenges to span the roots.
+        uint challengeCount = 7;
+        PDPService.Proof[] memory proofs = buildProofs(setId, challengeCount, trees, leafCounts);
+        // Submit proof.
+        pdpService.provePossession(setId, proofs);
+        // Verify the next challenge is in a subsequent epoch.
+        assertEq(pdpService.getNextChallengeEpoch(setId), block.number + challengeFinalityDelay);
+    }
+
+    // Builds a Merkle tree over data that is a sequence of distinct leaf values.
+    function makeTree(uint leafCount) internal view returns (bytes32[][] memory) {
+        bytes32[] memory data = ProofUtil.generateLeaves(leafCount);
+        bytes32[][] memory tree = MerkleProve.buildMerkleTree(data);
+        return tree;
+    }
+
+    // Constructs a RootData structure for a Merkle tree.
+    function makeRoot(bytes32[][] memory tree, uint leafCount) internal pure returns (PDPService.RootData memory) {
+        return PDPService.RootData(Cids.cidFromDigest(bytes(cidPrefix), tree[0][0]), leafCount * 32);
+    }
+
+    // Builds a proof of posession for a proof set
+    function buildProofs(uint256 setId, uint challengeCount, bytes32[][][] memory trees, uint[] memory leafCounts) internal view returns (PDPService.Proof[] memory) {
+        uint256 challengeEpoch = pdpService.getNextChallengeEpoch(setId);
+        uint256 seed = challengeEpoch; // Seed is (temporarily) the challenge epoch
+        uint totalLeafCount = 0;
+        for (uint i = 0; i < leafCounts.length; ++i) {
+            totalLeafCount += leafCounts[i];
+        }
+        
+        // console.log("seed", vm.toString(seed));
+        PDPService.Proof[] memory proofs = new PDPService.Proof[](challengeCount);
+        for (uint challengeIdx = 0; challengeIdx < challengeCount; challengeIdx++) {
+            // Compute challenge index
+            bytes memory payload = abi.encodePacked(seed, setId, uint64(challengeIdx));
+            uint256 challengeOffset = uint256(keccak256(payload)) % totalLeafCount;
+            // console.log("challengeIdx", vm.toString(challengeIdx));
+
+            uint treeIdx = 0;
+            uint256 treeOffset = 0;
+            for (uint i = 0; i < leafCounts.length; ++i) {
+                if (leafCounts[i] >= challengeOffset) {
```

Comment:
It doesn't look like this is the bug causing failure but I think this <= is off by one and should be <.  If we got unlucky and we landed with a challenge offset == leafCount[i], i.e. a leaf count of 5 and a challenge offset of 5 then we'd try to access leafs[5] which would overrun the leaf buffer.